services/ifrs/src/ifrs/migration_logic/_manager.py

```diff
@@ -0,0 +1,384 @@
+
+        Raises `MigrationError` if unsuccessful.
+        """
+        seq = self._get_sequence(current_ver=current_ver, target_ver=self.target_ver)
+        migrations = [self._fetch_migration_cls(ver) for ver in seq]
+        unapplying = self._migration_type == "BACKWARD"
+        try:
+            # Execute & time each migration in order to get to the target DB version
+            for v, migration_cls in zip(seq, migrations, strict=True):
+                # Determine if this is the last migration to apply/unapply
+                last_ver_called = self.target_ver + 1 if unapplying else self.target_ver
+                is_final_migration = v == last_ver_called
+
+                # instantiate MigrationDefinition
+                migration = migration_cls(
+                    db=self.db,
+                    unapplying=unapplying,
+                    is_final_migration=is_final_migration,
+                )
+
+                # Call apply/unapply based on migration type
+                await migration.unapply() if unapplying else await migration.apply()
+        except BaseException as exc:
+            error = MigrationStepError(
+                current_ver=current_ver, target_ver=self.target_ver, err_info=str(exc)
+            )
+            log.critical(error)
+            raise error from exc
+
+    async def _migrate_db(self) -> bool:
+        """Ensure the database is up to date before running the actual app.
+
+        If the database is already up to date, no changes are made. If the database is
+        out of date, migration code is executed to make the database current.
+
+        Returns True if migrations are finished or up-to-date and False otherwise.
+        """
+        version_docs = await self._get_version_docs()
+        version = _get_db_version_from_records(version_docs)
+
+        if version == 0:
+            try:
+                init_complete = await self._initialize_versioning()
+            except BaseException as exc:
+                error = DbVersioningInitError(err_info=str(exc))
+                log.critical(error)
+                raise error from exc
+            if not init_complete:
+                return False
+            version = 1
+
+        if version == self.target_ver:
+            # DB is up to date, run service
+            return True
+
+        # DB version is not what it should be: acquire lock and migrate
+        async with self._lock_db():
+            if not self._lock_acquired:
+                return False
+
+            self._migration_type = (
+                "FORWARD" if version < self.target_ver else "BACKWARD"
+            )
+
+            start = time()
+            await self._perform_migrations(current_ver=version)
+            duration_ms = duration_in_ms(time() - start)
+
+            # record the db version
+            await self._record_migration(
+                version=self.target_ver,
+                total_duration_ms=duration_ms,
+            )
```

Comment:
This means there is only one migration record, even if multiple versions are involved and only the target version is recorded?